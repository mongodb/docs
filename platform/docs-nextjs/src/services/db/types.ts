import { WithId } from "mongodb";

interface StaticAsset {
  checksum: string;
  key: string;
}

interface UpdatedAsset extends StaticAsset {
  updated_at?: Date;
}

interface PageAst {
  options?: {
    template?: string;
  }
  [key: string]: unknown;
}

interface Facet {
  category: string;
  value: string;
  display_name: string;
  sub_facets?: { [key: string]: unknown }[];
}

export interface ASTDocument extends WithId<Document> {
  page_id: string;
  page_path: string;
  filename: string;
  ast: PageAst;
  static_assets: UpdatedAsset[];
  facets?: Facet[];
}
