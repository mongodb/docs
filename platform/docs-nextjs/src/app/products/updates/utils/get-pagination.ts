/**
 *
 * @param array - flatten array
 * @param chunkSize - max size of each array
 * @returns [[{}, {}], [{}]]
 * @description Takes an flatten array and creates one level nested array of objects based on the chunkSize. Used for pagination data.
 */
export const getPagination = <T>(array: T[], chunkSize: number): T[][] => {
  const chunks: T[][] = [];
  if (!array.length) {
    return [[]]; // default to one empty nested array
  }
  for (let i = 0; i < array.length; i += chunkSize) {
    chunks.push(array.slice(i, i + chunkSize));
  }
  return chunks;
};
