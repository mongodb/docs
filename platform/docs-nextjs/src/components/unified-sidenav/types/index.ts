export interface BreadCrumb {
  path: string;
  title: string;
}

export interface TocItem {
  label: string;
  url?: string;
  newUrl?: string;
  group?: boolean;
  contentSite?: string;
  collapsible?: boolean;
  breadcrumbs?: BreadCrumb[];
  showSubNav?: boolean;
  isExternal?: boolean;
  versionDropdown?: boolean;
  versions?: {
    includes?: string[];
    excludes?: string[];
  };
  items?: TocItem[];
}
