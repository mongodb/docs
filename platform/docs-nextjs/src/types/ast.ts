// Used for Roles - Highlight
export const HIGHLIGHT_BLUE = 'highlight-blue';
export const HIGHLIGHT_GREEN = 'highlight-green';
export const HIGHLIGHT_RED = 'highlight-red';
export const HIGHLIGHT_YELLOW = 'highlight-yellow';

// Used for tab-context
export type Selectors = Record<string, Record<string, ASTNode[]>>;
export type ActiveTabs = Record<string, string>;

// Used for Page Templates
export type PageTemplateType =
  | 'blank'
  | 'drivers-index'
  | 'document'
  | 'errorpage'
  | 'feature-not-avail'
  | 'instruqt'
  | 'landing'
  | 'openapi'
  | 'changelog'
  | 'search'
  | 'guide'
  | 'product-landing';

type ComponentType =
  | Exclude<NodeType, 'directive' | 'directive_argument' | 'role' | 'target_identifier' | 'inline_target'>
  | 'admonition'
  | 'banner'
  | 'blockquote'
  | 'button'
  | 'card'
  | 'chapter'
  | 'chapters'
  | 'collapsible'
  | 'community-driver'
  | 'composable-tutorial'
  | 'cond'
  | 'container'
  | 'cta'
  | 'deprecated'
  | 'deprecated-version-selector'
  | 'describe'
  | 'extract'
  | 'field'
  | 'field_list'
  | 'figure'
  | 'footnote'
  | 'glossary'
  | 'hlist'
  | 'image'
  | 'include'
  | 'introduction'
  | 'io-code-block'
  | 'kicker'
  | 'literal_block'
  | 'listItem'
  | 'list'
  | 'list-table'
  | 'literalinclude'
  | 'openapi-changelog'
  | 'procedure'
  | 'ref_role'
  | 'role'
  | 'release_specification'
  | 'rubric'
  | 'search-results'
  | 'section'
  | 'seealso'
  | 'selected-content'
  | 'sharedinclude'
  | 'step'
  | 'substitution_reference'
  | 'tab'
  | 'tabs-selector'
  | 'time'
  | 'title_reference'
  | 'transition'
  | 'versionadded'
  | 'versionchanged'
  | 'tabs'
  | 'wayfinding';

type DirectiveName =
  | AdmonitionName
  | 'admonition'
  | 'banner'
  | 'blockquote'
  | 'card'
  | 'card-group'
  | 'chapter'
  | 'chapters'
  | 'collapsible'
  | 'community-driver'
  | 'composable-tutorials'
  | 'container'
  | 'contents'
  | 'deprecated'
  | 'directive'
  | 'dismissible-skills-card'
  | 'entry'
  | 'facet'
  | 'guide-next'
  | 'hlist'
  | 'figure'
  | 'ia'
  | 'icon'
  | 'image'
  | 'include'
  | 'input'
  | 'io-code-block'
  | 'list-table'
  | 'literalinclude'
  | 'meta'
  | 'openapi-changelog'
  | 'output'
  | 'procedure'
  | 'ref_role'
  | 'role'
  | 'release_specification'
  | 'rubric'
  | 'search-results'
  | 'section'
  | 'seealso'
  | 'selected-content'
  | 'sharedinclude'
  | 'step'
  | 'substitution_reference'
  | 'tab'
  | 'tabs-selector'
  | 'time'
  | 'transition'
  | 'toctree'
  | 'versionadded'
  | 'versionchanged'
  | 'tabs'
  | 'wayfinding'
  | 'wayfinding-option'
  | 'wayfinding-description';

type NodeType =
  | 'card-group'
  | 'code'
  | 'cta-banner'
  | 'definitionList'
  | 'definitionListItem'
  | 'directive'
  | 'directive_argument'
  | 'emphasis'
  | 'field'
  | 'field_list'
  | 'footnote'
  | 'footnote_reference'
  | 'guide-next'
  | 'heading'
  | 'inline_target'
  | 'line'
  | 'line_block'
  | 'list'
  | 'listItem'
  | 'literal'
  | 'method-selector'
  | 'only'
  | 'paragraph'
  | 'reference'
  | 'ref_role'
  | 'release_specification'
  | 'role'
  | 'root'
  | 'section'
  | 'strong'
  | 'superscript'
  | 'subscript'
  | 'tabs'
  | 'target'
  | 'target_identifier'
  | 'text'
  | 'title_reference';

type RoleName = (typeof roleNames)[number];

export const roleNames = [
  'abbr',
  'class',
  'command',
  'file',
  'guilabel',
  'icon',
  HIGHLIGHT_BLUE,
  HIGHLIGHT_GREEN,
  HIGHLIGHT_RED,
  HIGHLIGHT_YELLOW,
  'icon-fa5',
  'icon-fa5-brands',
  'icon-fa4',
  'icon-mms',
  'icon-charts',
  'icon-lg',
  'kbd',
  'red',
  'gold',
  'required',
  'sub',
  'subscript',
  'sup',
  'superscript',
  'link-new-tab',
];

type NodeName = RoleName | DirectiveName | AdmonitionName;

type DirectiveOptions = {
  [key: string]: string;
};

interface Node {
  type: NodeType;
  position: {
    start: {
      line: number | { $numberInt: string };
    };
  };
}

interface TextParentNode extends Node {
  children: TextNode[];
}

interface ParentNode extends Node {
  children: ASTNode[];
}

type PageOptionsKey = keyof PageOptions;

type HeadingOption = {
  depth: number;
  id: string;
  selector_ids: HeadingNodeSelectorIds;
  title: TextNode[];
};

type PageOptions = {
  template: PageTemplateType;
  default_tabs?: ActiveTabs;
  dismissible_skills_card?: DismissibleSkillsCardOptions;
  has_composable_tutorial?: boolean;
  hidefeedback?: string;
  instruqt?: boolean;
  has_method_selector?: boolean;
  multi_page_tutorial_settings?: {
    show_next_top?: boolean;
  };
  noprevnext?: string;
  'pl-max-width-paragraphs'?: string;
  selectors?: Selectors;
  'tabs-selector-position'?: string;
  time_required?: number;
  headings?: HeadingNode[];
};

interface Root extends ParentNode {
  type: 'root';
  options?: PageOptions;
  fileid: string;
}

interface ChaptersNode extends Directive {
  name: 'chapters';
}

type ChapterOptions = {
  description: string;
  image: string;
};

interface ChapterNode extends Directive<ChapterOptions> {
  name: 'chapter';
}

interface FootnoteReferenceNode extends ParentNode {
  type: 'footnote_reference';
  id: string;
  refname?: string;
}

interface FootnoteNode extends ParentNode {
  type: `footnote`;
  id: string;
  name?: string;
}

type HeadingNodeSelectorIds = {
  tab?: string;
  'method-option'?: string;
  'selected-content'?: Record<string, string>;
  children?: HeadingNodeSelectorIds;
};

interface HeadingNode extends ParentNode {
  type: 'heading';
  depth: number;
  title: string;
  id: string;
  selector_ids: HeadingNodeSelectorIds;
  children: [TextNode];
}

interface ParagraphNode extends ParentNode {
  type: 'paragraph';
}

interface EmphasisNode extends ParentNode {
  type: 'emphasis';
  children: [TextNode];
}

interface StrongNode extends TextParentNode {
  type: 'strong';
}

interface SuperscriptNode extends ParentNode {
  type: 'superscript';
}

interface SubscriptNode extends ParentNode {
  type: 'subscript';
}

interface ReferenceNode extends ParentNode {
  type: 'reference';
  refuri: string;
}

interface Directive<TOptions = DirectiveOptions> extends ParentNode {
  type: 'directive';
  name: DirectiveName;
  argument: ASTNode[];
  domain?: string;
  options?: TOptions;
}

interface BlockQuoteNode extends Directive {
  name: 'blockquote';
}

type ButtonOptions = {
  uri: string;
};

interface ButtonNode extends Directive<ButtonOptions> {
  options: ButtonOptions;
}

interface ContainerNode extends Directive {
  name: 'container';
  argument: TextNode[];
}

type DismissibleSkillsCardOptions = {
  skill: string;
  url: string;
};

interface DismissibleSkillsCardNode extends Directive<DismissibleSkillsCardOptions> {
  name: 'dismissible-skills-card';
  options: DismissibleSkillsCardOptions;
}

type ListTableOptions = {
  align?: string;
  width?: string;
  widths?: string;
  'header-rows'?: string;
  'stub-columns'?: string;
};

interface ListTableNode extends Directive<ListTableOptions> {
  name: 'list-table';
  children: ParentListNode[];
  options?: ListTableOptions;
}

interface BaseFieldNode extends ParentNode {
  name: string;
  label?: string;
}

interface FieldNode extends BaseFieldNode {
  type: 'field';
}

interface FieldListNode extends BaseFieldNode {
  type: 'field_list';
  children: FieldNode[];
}

interface ParentListNode extends ParentNode {
  type: 'list';
  enumtype?: 'unordered' | 'ordered';
  startat?: number;
  children: ParentListItemNode[];
}

interface ListNode extends ParentNode {
  type: 'list';
  enumtype?: 'unordered' | 'ordered' | '1' | 'a' | 'A' | 'i' | 'I';
  startat?: number;
  children: ListItemNode[];
}

interface ParentListItemNode extends ParentNode {
  type: 'listItem';
  children: ListNode[];
}

interface ListItemNode extends ParentNode {
  type: 'listItem';
  children: ParentNode[];
}

interface LiteralNode extends ParentNode {
  type: 'literal';
}

interface LineBlockNode extends ParentNode {
  type: 'line_block';
}

interface LineNode extends ParentNode {
  type: 'line';
}

interface TextNode extends Node {
  type: 'text';
  value: string;
}

type CardGroupOptions = {
  columns: number;
  layout?: string;
  style?: string;
  type?: string;
};

interface CardGroupNode extends Directive<CardGroupOptions> {
  name: 'card-group';
  options: CardGroupOptions;
  children: CardNode[];
}

type CardOptions = {
  cta?: string;
  headline?: string;
  icon?: string;
  'icon-dark'?: string;
  'icon-alt'?: string;
  tag?: string;
  url?: string;
};

interface CardNode extends Directive<CardOptions> {
  name: 'card';
  options: CardOptions;
}

interface DefinitionListNode extends ParentNode {
  type: 'definitionList';
  children: DefinitionListItemNode[];
}

interface DefinitionListItemNode extends ParentNode {
  type: 'definitionListItem';
  term: ASTNode[];
}

interface CodeNode extends Node {
  type: 'code';
  lang: string;
  copyable: boolean;
  emphasize_lines: number[];
  value: string;
  linenos: boolean;
  caption?: string;
  source?: string;
  lineno_start?: number;
}

type IOCodeBlockOptions = {
  copyable: boolean;
};

interface IOCodeBlockNode extends Directive<IOCodeBlockOptions> {
  name: 'io-code-block';
  children: [IOInputNode] | [IOInputNode, IOOutputNode];
  options: IOCodeBlockOptions;
}

type InputOutputOptions = {
  language: string;
  linenos: boolean;
  visible?: boolean;
};

interface IOInputNode extends Directive<InputOutputOptions> {
  name: 'input';
  children: CodeNode[];
  options: InputOutputOptions;
}

interface IOOutputNode extends Directive<InputOutputOptions> {
  name: 'output';
  children: CodeNode[];
  options: InputOutputOptions;
}

interface MethodNode extends ParentNode {
  type: 'target';
  name: 'method';
  domain: 'mongodb';
}

interface DirectiveArgumentNode extends ParentNode {
  type: 'directive_argument';
}

interface BaseTargetNode extends ParentNode {
  domain: string;
  name: string;
  html_id?: string;
  options?: object;
}

interface TargetNode extends BaseTargetNode {
  type: 'target';
}

interface InlineTargetNode extends BaseTargetNode {
  type: 'inline_target';
}

interface TargetIdentifierNode extends ParentNode {
  type: 'target_identifier';
  ids: string[];
}

interface AbbrRoleNode extends ParentNode {
  type: 'role';
  name: 'abbr';
  children: [TextNode];
}

type CollapsibleOptions = {
  heading?: string;
  sub_heading?: string;
  id?: string;
  expanded?: boolean;
};

interface CollapsibleNode extends Directive<CollapsibleOptions> {
  name: 'collapsible';
  children: ASTNode[];
  options?: CollapsibleOptions;
}

interface ContentsOptions {
  local?: boolean;
  backlinks?: string;
  depth?: number;
  class?: string;
}

interface ContentsNode extends Directive<ContentsOptions> {
  name: 'contents';
  options: ContentsOptions;
}

interface TabsNode extends Directive<TabsOptions> {
  name: 'tabs';
  children: TabNode[];
}

interface TabsOptions {
  tabset?: string;
  hidden?: boolean;
}

interface TabOptions {
  tabid: string;
}

interface TabNode extends Directive<TabOptions> {
  type: 'directive';
  name: 'tab';
  options?: TabOptions;
}

interface FacetOptions {
  name: string;
  values: string;
}

interface FacetNode extends Directive<FacetOptions> {
  type: 'directive';
  name: 'facet';
  options?: FacetOptions;
}

type ImageNodeOptions = {
  alt: string;
  lightbox?: boolean;
  align?: string;
  checksum?: string;
  height?: string;
  scale?: string;
  width?: string;
  figwidth?: string;
  loading?: string;
  class?: string;
  border?: boolean;
};

interface FigureNode extends Directive<ImageNodeOptions> {
  name: 'figure';
  argument: [TextNode];
  options: ImageNodeOptions;
}

interface ImageNode extends Directive<ImageNodeOptions> {
  name: 'image';
  argument: [TextNode];
  options: ImageNodeOptions;
}

type AdmonitionName = 'example' | 'note' | 'tip' | 'see' | 'seealso' | 'warning' | 'important';

interface AdmonitionNode extends Directive {
  type: 'directive';
  name: AdmonitionName;
}

type TocTreeOptions = {
  drawer?: boolean;
  project?: string;
  versions?: string[];
  osiris_parent?: boolean;
  tocicon?: string;
  version?: string;
  urls?: Record<string, string>;
};
interface TocTreeEntry extends ParentNode {
  title: [TextNode];
  slug: string;
  children: TocTreeEntry[];
  options?: TocTreeOptions;
  url?: string;
}

interface TocTreeDirective extends Directive<TocTreeOptions> {
  type: 'directive';
  name: 'toctree';
  entries: Array<TocTreeEntry>;
}

type BannerOptions = {
  variant: 'info' | 'warning' | 'danger';
};

interface BannerNode extends Directive<BannerOptions> {
  options: BannerOptions;
}

type CTABannerOptions = {
  url: string;
  icon?: string;
};

interface CTABannerNode extends Directive<CTABannerOptions> {
  options: CTABannerOptions;
}

interface ClassRoleNode extends ParentNode {
  type: 'role';
  name: 'class';
  target: string;
}

type CommunityDriverPillOptions = {
  url: string;
};

interface CommunityDriverPill extends Directive<CommunityDriverPillOptions> {
  name: 'community-driver';
  options: CommunityDriverPillOptions;
}

interface ComposableTutorialOption {
  default: string;
  dependencies: Record<string, string>[];
  // selections used to display list of dropdowns
  // ie. [{value: 'nodejs', text: 'Node.js'}, {value: 'cpp', text: 'C++'}]
  selections: { value: string; text: string }[];
  text: string;
  value: string;
}

interface ComposableTutorialNode extends Directive {
  type: 'directive';
  name: 'composable-tutorials';
  children: ComposableNode[];
  composable_options: ComposableTutorialOption[];
}

interface ComposableNode extends Directive {
  type: 'directive';
  name: 'selected-content';
  // selections required to show this composable node
  // ie. {interface: 'drivers', language: 'nodejs'}
  selections: Record<string, string>;
  children: ASTNode[];
}

const highlightRoleNames = [HIGHLIGHT_BLUE, HIGHLIGHT_GREEN, HIGHLIGHT_RED, HIGHLIGHT_YELLOW];
type HighlightRoleNames = (typeof highlightRoleNames)[number];

interface HighlightNode extends ParentNode {
  type: 'role';
  name: HighlightRoleNames;
}

interface IANode extends Directive {
  name: 'ia';
}

type IAEntryNodeOptions = {
  url: string;
};

interface IAEntryNode extends Directive {
  name: 'entry';
  options: IAEntryNodeOptions;
}

interface LinkNewTabNode extends ParentNode {
  type: 'role';
  name: 'link-new-tab';
  target: string;
}

const roleIconNames = [
  'icon',
  'icon-fa5-brands',
  'iconb',
  'icon-mms',
  'icon-mms-org',
  'icon-charts',
  'icon-fa4',
  'icon-lg',
];
type RoleIconNames = (typeof roleIconNames)[number];

interface RoleIconNode extends ParentNode {
  type: 'role';
  name: RoleIconNames;
  target: string;
}

interface RoleManualNode extends ParentNode {
  type: 'role';
  name: 'manual';
  target: string;
}

type MetaOptions = {
  canonical: string;
  description?: string;
  robots?: string;
  keywords?: string;
};

interface MetaNode extends Directive<MetaOptions> {
  type: 'directive';
  name: 'meta';
  options: MetaOptions;
}

type ProcedureStyle = 'connected' | 'normal';

type ProcedureOptions = {
  style?: ProcedureStyle;
  title?: string;
};

interface ProcedureNode extends Directive<ProcedureOptions> {
  name: 'procedure';
  options: ProcedureOptions;
}

interface StepNode extends Directive {
  name: 'step';
}

interface TitleReferenceNode extends ParentNode {
  type: 'title_reference';
  children: TextNode[];
}

type TwitterOptions = {
  creator?: string;
  image?: string;
  'image-alt'?: string;
  site?: string;
  title?: string;
};

interface TwitterNode extends Directive<TwitterOptions> {
  options: TwitterOptions;
}

type HorizontalListNodeOptions = {
  columns: number;
};

interface HorizontalListNode extends Directive<HorizontalListNodeOptions> {
  name: 'hlist';
  children: [ListNode];
  options: HorizontalListNodeOptions;
}

type VideoOptions = {
  title: string;
  description: string;
  'upload-date': string;
  'thumbnail-url': string;
};

interface VideoNode extends Directive<VideoOptions> {
  argument: ReferenceNode[];
}

interface WayfindingDescriptionNode extends Directive {
  name: 'wayfinding-description';
}

interface WayfindingNode extends Directive {
  argument: ReferenceNode[];
}

type WayfindingOptionOptions = {
  title: string;
  language: string;
  id: string;
};

interface WayfindingOptionNode extends Directive<WayfindingOptionOptions> {
  argument: ReferenceNode[];
}

type StandaloneHeaderOptions = {
  columns: number;
  cta: string;
  url: string;
};

interface StandaloneHeaderNode extends Directive<StandaloneHeaderOptions> {
  options: StandaloneHeaderOptions;
}

interface SubstitutionReferenceNode extends ParentNode {
  name: 'substitution_reference';
}

interface TargetNode extends ParentNode {
  name: 'target';
  html_id: string;
  options: {
    hidden: boolean;
  };
}

interface ReleaseSpecificationNode extends ParentNode {
  type: 'release_specification';
}

interface RefRoleNode extends ParentNode {
  type: 'ref_role';
  name: string;
  domain: string;
  target: string;
  url: string;
  fileid?: string[];
  flag?: string;
}

type InstruqtOptions = {
  title: string;
  drawer?: boolean;
};

interface InstruqtNode extends Directive<InstruqtOptions> {
  argument: TextNode[];
}

interface GuideNextNode extends Directive {
  name: 'guide-next';
}

type ASTNode =
  | AbbrRoleNode
  | AdmonitionNode
  | BannerNode
  | BlockQuoteNode
  | ButtonNode
  | CardGroupNode
  | CardNode
  | ChapterNode
  | ChaptersNode
  | ClassRoleNode
  | CodeNode
  | CollapsibleNode
  | CommunityDriverPill
  | ContentsNode
  | ComposableNode
  | ComposableTutorialNode
  | ContainerNode
  | CTABannerNode
  | DefinitionListNode
  | DefinitionListItemNode
  | Directive
  | DirectiveArgumentNode
  | DismissibleSkillsCardNode
  | EmphasisNode
  | FacetNode
  | FieldNode
  | FieldListNode
  | FigureNode
  | FootnoteNode
  | FootnoteReferenceNode
  | GuideNextNode
  | HeadingNode
  | HighlightNode
  | ImageNode
  | InlineTargetNode
  | InstruqtNode
  | IOCodeBlockNode
  | IOInputNode
  | IOOutputNode
  | LinkNewTabNode
  | ListNode
  | ListTableNode
  | ListItemNode
  | LiteralNode
  | LineBlockNode
  | LineNode
  | MetaNode
  | MethodNode
  | ParagraphNode
  | ParentNode
  | ProcedureNode
  | ReferenceNode
  | StandaloneHeaderNode
  | SuperscriptNode
  | SubscriptNode
  | RefRoleNode
  | ReleaseSpecificationNode
  | RoleIconNode
  | RoleManualNode
  | Root
  | StepNode
  | StrongNode
  | SubstitutionReferenceNode
  | TabNode
  | TabsNode
  | TargetIdentifierNode
  | TargetNode
  | TextNode
  | TitleReferenceNode
  | TwitterNode
  | TocTreeEntry
  | TocTreeDirective
  | VideoNode
  | WayfindingDescriptionNode
  | WayfindingNode
  | WayfindingOptionNode;

export { highlightRoleNames, roleIconNames };

export type {
  AbbrRoleNode,
  AdmonitionNode,
  AdmonitionName,
  ASTNode,
  BannerNode,
  BlockQuoteNode,
  ButtonNode,
  CardGroupNode,
  CardNode,
  ChapterNode,
  ChaptersNode,
  ClassRoleNode,
  CodeNode,
  CollapsibleNode,
  CollapsibleOptions,
  CommunityDriverPill,
  ContentsNode,
  ComponentType,
  ComposableNode,
  ComposableTutorialNode,
  ComposableTutorialOption,
  ContainerNode,
  CTABannerNode,
  DefinitionListNode,
  DefinitionListItemNode,
  Directive,
  DirectiveArgumentNode,
  DirectiveName,
  DirectiveOptions,
  DismissibleSkillsCardNode,
  EmphasisNode,
  FacetNode,
  FieldNode,
  FieldListNode,
  FigureNode,
  FootnoteNode,
  FootnoteReferenceNode,
  GuideNextNode,
  HeadingNode,
  HeadingNodeSelectorIds,
  HeadingOption,
  HighlightNode,
  HighlightRoleNames,
  HorizontalListNode,
  IAEntryNode,
  IANode,
  ImageNode,
  ImageNodeOptions,
  InlineTargetNode,
  InstruqtNode,
  IOCodeBlockNode,
  IOInputNode,
  IOOutputNode,
  LinkNewTabNode,
  ListNode,
  ListTableNode,
  ListItemNode,
  LiteralNode,
  LineBlockNode,
  LineNode,
  MetaNode,
  MethodNode,
  Node,
  NodeName,
  NodeType,
  PageOptions,
  PageOptionsKey,
  ParagraphNode,
  ParentListItemNode,
  ParentListNode,
  ParentNode,
  ProcedureNode,
  ProcedureStyle,
  ReferenceNode,
  StandaloneHeaderNode,
  SuperscriptNode,
  SubscriptNode,
  RefRoleNode,
  ReleaseSpecificationNode,
  RoleIconNode,
  RoleManualNode,
  RoleName,
  Root,
  StepNode,
  StrongNode,
  SubstitutionReferenceNode,
  TabNode,
  TabsNode,
  TabsOptions,
  TargetIdentifierNode,
  TargetNode,
  TextNode,
  TitleReferenceNode,
  TwitterNode,
  TocTreeEntry,
  TocTreeDirective,
  VideoNode,
  WayfindingDescriptionNode,
  WayfindingNode,
  WayfindingOptionNode,
};
