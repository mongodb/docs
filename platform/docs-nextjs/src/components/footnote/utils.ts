import type { ASTNode } from '@/types/ast';
import type { FootnoteNode, FootnoteReferenceNode } from '@/types/ast';
import { findAllKeyValuePairs } from '@/utils/find-all-key-value-pairs';
import type { FootnoteType } from './footnote-context';

export const getFootnotes = (nodes: ASTNode[]) => {
  const footnotes = findAllKeyValuePairs(nodes, 'type', 'footnote') as FootnoteNode[];
  const footnoteReferences = findAllKeyValuePairs(nodes, 'type', 'footnote_reference') as FootnoteReferenceNode[];
  const numAnonRefs = footnoteReferences.filter(
    (node) => !Object.prototype.hasOwnProperty.call(node, 'refname'),
  ).length;
  // We label our footnotes by their index, regardless of their names to
  // circumvent cases such as [[1], [#], [2], ...]
  return footnotes.reduce<Record<string, FootnoteType>>((map, footnote, index) => {
    if (footnote.name) {
      // Find references associated with a named footnote

      map[footnote.name] = {
        label: index + 1,
        references: getNamedFootnoteReferences(footnoteReferences, footnote.name),
      };
    } else {
      // Find references associated with an anonymous footnote
      // Replace potentially broken anonymous footnote ids
      footnote.id = `${index + 1}`;

      map[footnote.id] = {
        label: index + 1,
        references: getAnonymousFootnoteReferences(index, numAnonRefs),
      };
    }
    return map;
  }, {});
};

// Find all footnote_reference node IDs associated with a given footnote by
// that footnote's refname
const getNamedFootnoteReferences = (footnoteReferences: FootnoteReferenceNode[], refname: string) => {
  return footnoteReferences.filter((node) => node.refname === refname).map((node) => node.id);
};

// They are used infrequently, but here we match an anonymous footnote to its reference.
// The nth footnote on a page is associated with the nth reference on the page. Since
// anon footnotes and footnote references are anonymous, we assume a 1:1 pairing, and
// have no need to query nodes. If there are more anonymous footnotes than references,
// we may return an empty array
const getAnonymousFootnoteReferences = (index: number, numAnonRefs: number) => {
  return index > numAnonRefs ? [] : [`id${index + 1}`];
};
