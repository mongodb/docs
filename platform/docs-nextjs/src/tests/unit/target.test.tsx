import React from 'react';
import { render } from '@testing-library/react';
import Target from '@/components/target';
import { mockLocation } from '@/tests/utils/mock-location';

// data for this component
import mockData from '@/tests/data/target.test.json';
import { ASTNode } from '@/types/ast';

beforeAll(() => {
  mockLocation({hash: ''});
});

it('renders correctly with a directive_argument node', () => {
  const tree = render(<Target nodeChildren={mockData.a.children as ASTNode[]} html_id={mockData.a.html_id} name={mockData.a.name}/>);
  expect(tree.asFragment()).toMatchSnapshot();
});

it('renders correctly with no directive_argument nodes', () => {
  const tree = render(<Target nodeChildren={mockData.b.children as ASTNode[]} html_id={mockData.b.html_id} name={mockData.b.name}/>);
  expect(tree.asFragment()).toMatchSnapshot();
});
