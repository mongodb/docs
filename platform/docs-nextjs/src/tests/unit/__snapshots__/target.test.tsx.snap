// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`renders correctly with a directive_argument node 1`] = `
<DocumentFragment>
  .emotion-0 {
  font-size: 13px;
  line-height: 20px;
  display: inline;
  -webkit-transition: all 0.15s ease-in-out;
  transition: all 0.15s ease-in-out;
  border-radius: 3px;
  font-family: 'Source Code Pro',Menlo,monospace;
  line-height: 20px;
  background-color: #F9FBFA;
  border: 1px solid #E8EDEB;
  color: #1C2D38;
  white-space: nowrap;
  font-size: unset;
  display: inline;
  color: var(--font-color-primary);
  background: var(--background-color-secondary);
  word-wrap: break-word;
  white-space: unset;
}

.lg-ui-0000:hover>.emotion-0 {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.lg-ui-0000:hover>.emotion-0 {
  box-shadow: 0 0 0 3px #E8EDEB;
  border: 1px solid #C1C7C6;
}

.lg-ui-0000:focus-visible>.emotion-0 {
  box-shadow: 0 0 0 2px #FFFFFF,0 0 0 4px #0498EC;
  border: 1px solid #016BF8;
}

a .emotion-0 {
  color: inherit;
}

.emotion-1 {
  position: absolute;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  padding: 0 10px;
  visibility: hidden;
}

.emotion-2 {
  color: #889397;
  vertical-align: middle;
  margin-top: -2px;
}

.emotion-3 {
  display: none;
}

.emotion-3 {
  display: none;
}

.emotion-4 {
  display: inline;
  left: 0;
  top: 0;
  margin-top: -85px;
  position: absolute;
  padding-bottom: 2px;
}

@media only screen and (max-width: 767px) {
  .emotion-4 {
    margin-top: -85px;
  }
}

.emotion-5 {
  margin: unset;
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  color: #001E2B;
  font-size: 13px;
  line-height: 20px;
  color: #001E2B;
  font-weight: 400;
  margin-bottom: 16px;
  color: var(--font-color-primary);
}

.emotion-5 strong,
.emotion-5 b {
  font-weight: 600;
}

<dl
    class="option"
  >
    <dt>
      <code
        class="emotion-0"
      >
        --config &lt;filename&gt;, -f &lt;filename&gt;
      </code>
      <a
        class="headerlink emotion-1"
        href="#std-option-mongos.--config"
        title="Permalink to this definition"
      >
        <svg
          aria-label="Link Icon"
          class="emotion-2"
          fill="none"
          height="12"
          role="img"
          viewBox="0 0 16 16"
          width="12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6.039 9.953a4 4 0 0 1-.327-.38l1.45-1.445q.11.226.294.414a1.553 1.553 0 0 0 2.203.01l2.884-2.872a1.57 1.57 0 0 0 .01-2.212 1.553 1.553 0 0 0-2.203-.01l-.379.377A.995.995 0 0 1 8.56 3.83a1.005 1.005 0 0 1 .006-1.418l.38-.377a3.54 3.54 0 0 1 5.024.023 3.58 3.58 0 0 1-.022 5.047l-2.884 2.871a3.54 3.54 0 0 1-5.025-.022"
            fill="currentColor"
          />
          <path
            d="M9.961 6.047q.18.181.327.38l-1.45 1.445q-.11-.226-.294-.414a1.553 1.553 0 0 0-2.203-.01L3.457 10.32a1.57 1.57 0 0 0-.01 2.212 1.553 1.553 0 0 0 2.203.01l.379-.377a.995.995 0 0 1 1.411.006 1.005 1.005 0 0 1-.006 1.418l-.38.377a3.54 3.54 0 0 1-5.024-.023 3.58 3.58 0 0 1 .022-5.047l2.884-2.871a3.54 3.54 0 0 1 5.025.022"
            fill="currentColor"
          />
        </svg>
        <span
          class="emotion-3"
        />
        <div
          class="emotion-4"
          id="std-option-mongos.--config"
        />
      </a>
    </dt>
    <dd>
      <p
        class="emotion-5"
      >
        Specifies a configuration file for runtime configuration options. The
configuration file is the preferred method for runtime configuration of

        <span
          class="component-container"
        >
          Component for ref_role "program" not yet implemented
          <br />
          <span>
            <code
              class="emotion-0"
            >
              mongos
            </code>
          </span>
        </span>
        . The options are equivalent to the command-line
configuration options. See 
        <span
          class="component-container"
        >
          Component for ref_role "doc" not yet implemented
          <br />
          <span>
            Configuration File Options
          </span>
        </span>
         for
more information.
      </p>
      <p
        class="emotion-5"
      >
        Ensure the configuration file uses ASCII encoding. The 
        <span
          class="component-container"
        >
          Component for ref_role "program" not yet implemented
          <br />
          <span>
            <code
              class="emotion-0"
            >
              mongos
            </code>
          </span>
        </span>
        
instance does not support configuration files with non-ASCII encoding,
including UTF-8.
      </p>
    </dd>
  </dl>
</DocumentFragment>
`;

exports[`renders correctly with no directive_argument nodes 1`] = `
<DocumentFragment>
  .emotion-0 {
  scroll-margin-top: 85px;
  position: absolute;
}

<span
    class="emotion-0"
    id="std-label-php-language-center"
  />
</DocumentFragment>
`;
